stumpy/config.py

```diff
@@ -8,9 +8,11 @@
 STUMPY_MEAN_STD_NUM_CHUNKS = 1
 STUMPY_MEAN_STD_MAX_ITER = 10
 STUMPY_DENOM_THRESHOLD = 1e-14
-STUMPY_STDDEV_THRESHOLD = 1e-7
+STUMPY_STDDEV_THRESHOLD = 1e-20
```

Comment:
I don't think we should change this?

Reply:
> I don't think we should change this?

So, I think to answer this question, we need to answer the following (philosophical!) question first:

Let's say my data is:
```
seed = 0
T = np.random.rand(64)
identical = T[:8].copy() 
T[:8] = identical * 1e-7
T[-8:] = identical

# m is 8
```

Note that the z-norm of `T[:8]` and `T[-8:]` are the same. Now, which of the two statements below is correct?
* The z-norm Euclidean distance between `T[:8]` and `T[-8:]` is zero because their z-norm is the same.
* The z-norm Euclidean distance between `T[:8]` and `T[-8:]` is $\sqrt{m}$ (because `T[:8]` should be treated as a constant subsequence (The distance between a constant subsequence and a non-constant subsequence is $\sqrt{m}$ )

If we go with the latter statement, then `STUMPY_STDDEV_THRESHOLD = 1e-7` is fine. If we choose the former statement, then it should be okay to do:  `STUMPY_STDDEV_THRESHOLD = 1e-100`

Reply:
Going purely by my instincts, I think it should be the former (the distance should be zero because their z-norm are identical). 

So, it looks like we only use `STUMPY_STDDEV_THRESHOLD` in two functions:

https://github.com/TDAmeritrade/stumpy/blob/9a2e2ec522dbb47d14ba1bb3aba7515a9548cbae/stumpy/core.py#L923-L935

AND

https://github.com/TDAmeritrade/stumpy/blob/9a2e2ec522dbb47d14ba1bb3aba7515a9548cbae/stumpy/gpu_stump.py#L140-L155

In both cases, it feels like we could/should replace the use of that threshold with `T_subseq_isconstant` instead? Would that help at all?

Reply:
Okay, I think we are struggling with this because we are trying to make assumptions for the user (i.e., what should be considered "constant"). The purest/strictest definition is when the STDDEV is equal to zero and I doubt that there is any argument to this, right? Then, there is everything else that is an approximation.

I am starting to think if we should add `T_A_subseq_isconstant=None` and `T_B_subseq_isconstant=None` to our top level API which lets the user define the truth and we stop guessing. Any thoughts?

Reply:
> Going purely by my instincts, I think it should be the former (the distance should be zero because their z-norm are identical).

I have the same opinion.

> In the first two cases, it feels like we could/should replace the use of that threshold with `T_subseq_isconstant` instead?

Yes... I think it is better to use `T_subseq_isconstant`.



> However, in the 3rd case, we are using `STUMPY_STDDEV_THRESHOLD` to set `T_subseq_isconstant`
> is there a different/better way to identify a constant subsequence besides using the STDDEV?

If you are willing to sacrifice a little speed, then:
```
# rolling_min = calculate_rolling_min(T, m)  # NEW
T_is_constant = rolling_min == M_T 
```

> Just to state it out loud, it appears that it is not possible/trivial to distinguish whether a subsequence is constant OR whether it simply has very small values to start with OR both.

> Okay, I think we are struggling with this because we are trying to make assumptions for the user (i.e., what should be considered "constant"). The purest/strictest definition is when the STDDEV is equal to zero and I doubt that there is any argument to this, right? Then, there is everything else that is an approximation.

Exactly. The example I provided in my previous comment is a rare case. The user needs to decide what the right thing to do is. I mean... a user might see `T[:8] = identical * 1e-7` as a subsequence that is all zero but with some noise! If we plot the time series, we probably see this subsequence as a flat line. That is why I said this is somehow philosophical! I think we should treat timeseries `T` as is and it is better to not make any assumption for the user if possible. 



> I am starting to think if we should add `T_A_subseq_isconstant=None` and `T_B_subseq_isconstant=None` to our top level API which lets the user define the truth and we stop guessing. Any thoughts?

Yes. I think this is good! If it is not None, it is user's responsibility (we may still check the `std` values and throw out a warning if we notice the provided array for `T_A_subseq_isconstant` contains True for  a subseq that has `std>1`, for instance.)

If it is None, then we can either do `T_subseq_isconstant = T_std == 0` or use the `rolling_min` approach, which should contain less numerical error. 

If you are willing to sacrifice a little more speed, then this
```
# rolling_min = calculate_rolling_min(T, m)  # NEW
# rolling_max = calculate_rolling_max(T, m)  # NEW
T_is_constant = rolling_min == rolling_max
```
should be better as it has no numerical error.

Reply:
> > I am starting to think if we should add `T_A_subseq_isconstant=None` and `T_B_subseq_isconstant=None` to our top level API which lets the user define the truth and we stop guessing. Any thoughts?
> 
> Yes. I think this is good! If it is not None, it is user's responsibility (we may still check the `std` values and throw out a warning if we notice the provided array for `T_A_subseq_isconstant` contains True for a subseq that has `std>1`, for instance.)

How about `config.std_constant_threshold`? default is `0`. User can change it to 1e-3 or 1e-7 or whatever they want. 


```
If  `config.std_constant_threshold` != 0:
     T_is_constant = rolling_std <= `config.std_constant_threshold`
```

Reply:
> How about config.std_constant_threshold? 

I'd rather treat STDDEV == 0 as the only way to find a constant subsequence and then force the user to handle it on their own and then pass in `T_A_subseq_isconstant`. Doing it this way allows the most freedom. 

I think this should be a separate PR/issue?

Reply:
> I think we'd simply do:
> 
> ```
> # `stumpy.stump`
> 
> if T_A_subseq_isconstant is None:
>     T_A_subseq_isconstant = σ_Q < config.STUMPY_STDDEV_THRESHOLD
> 
> if T_B_subseq_isconstant is None:
>     T_B_subseq_isconstant = Σ_T < config.STUMPY_STDDEV_THRESHOLD
> ```
> 
> and then we'd use `T_A_subseq_isconstant` and `T_B_subseq_isconstant` to check whether the subsequence is constant. And then we'd leave it as `config.STUMPY_STDDEV_THRESHOLD = 1e-7` as the default and if the user doesn't like it then they either choose to change `config.STUMPY_STDDEV_THRESHOLD` to whatever they want OR pass in `T_A_subseq_isconstant` and/or `T_B_subseq_isconstant` directly.
> 
> What's really interesting is that if you allow the user to specify `T_A_subseq_isconstant` then they can choose to apply different constant rules to different subsequences. Imagine your example above:
> 
> ```
> seed = 0
> T = np.random.rand(64)
> identical = T[:8].copy() 
> T[:8] = identical * 1e-7
> T[-8:] = identical
> ```
> 
> The user can actually choose to treat `T[:8]` and `T[-8:]` completely independently:
> 
> ```
> T_subseq_isconstant = np.full(64 - m + 1, False)
> T_subseq_isconstant[0] = False
> T_subseq_isconstant[-1] = True
> ```
> 
> or any permutation of `True/False` for the first and last subsequences (or any other subsequence for that matter).

I can understand your point of view. I think, however, this might make things a little more complicated. I mean...it would not be easy for a user to set `T_subseq_isconstant`. Right? Because, they probably need to calculate the rolling std and then use it to compute `T_subseq_isconstant`. And, this may be used by only a few users. I know you are usually interested in solving the issues for general audience. But, if you think this is fine, then I think your proposed solution is a good :)

Reply:
> I mean...it would not be easy for a user to set T_subseq_isconstant. Right? Because, they probably need to calculate the rolling std and then use it to compute T_subseq_isconstant. And, this may be used by only a few users. I know you are usually interested in solving the issues for general audience. But, if you think this is fine, then I think your proposed solution is good :)

You raise a valid point. So, first things first, does setting `T_A_subseq_isconstant = σ_Q < config.STUMPY_STDDEV_THRESHOLD` and `T_B_subseq_isconstant = Σ_T < config.STUMPY_STDDEV_THRESHOLD` help with our constant subsequence precision issue? 

Then, separately, do we expose `T_A_subseq_isconstant` and `T_B_subseq_isconstant` to the user? It might make sense to take a moment and dig through our past issues to see what issues users were having when dealing with constant subsequences or near constant subsequences.

Again, I think the first part is more important while this second part is "nice to have" but, to your point, it's really for advanced users!

Reply:
> does setting `T_A_subseq_isconstant = σ_Q < config.STUMPY_STDDEV_THRESHOLD` and `T_B_subseq_isconstant = Σ_T < config.STUMPY_STDDEV_THRESHOLD` help with our constant subsequence precision issue?

I think we are (indirectly) doing it already:
* In `stumpy/stump.py`: we use `T_A_subseq_isconstant` which is one of the outputs returned by `core. preprocess_diagonal` in which we do `T_subseq_isconstant = Σ_T < config.STUMPY_STDDEV_THRESHOLD`.
* In `core._calculate_squared_distance`, we do the same. For instance, let's see lines 923-924 in `core.py`:

```
# In `core._calculate_squared_distance` (line 923)
if σ_Q < config.STUMPY_STDDEV_THRESHOLD or Σ_T < config.STUMPY_STDDEV_THRESHOLD:
            D_squared = m 
```
which is equivalent to:
```
# in `stumpy/stump.py` (line 185)
 if T_B_subseq_isconstant[i + k] or T_A_subseq_isconstant[i]:
         pearson = 0.5 # NOTE: this is equivalent to `D_squared = m`
```

---

btw, if I remember correctly, setting `config.STUMPY_STDDEV_THRESHOLD` to lower value helped me with resolving imprecision in one case. I cannot remember that particular case, but I can remember a test failed and the reason was the imprecision. The naive version gave `0` as the distance between the two identical case, but the performant version treated one of the identical subseq as constant and gave $\sqrt{m}$ as their distance. (similar to the example I provided earlier). That was the main reason behind setting `config.STUMPY_STDDEV_THRESHOLD=1e-20`. We can even set it lower, like `1e-100` but I am not sure if it makes sense or not.

---

> do we expose T_A_subseq_isconstant and T_B_subseq_isconstant to the user? 

**Did you mean if it is reasonable or not to allow user to set these two parameters?**

---

> It might make sense to take a moment and dig through our past issues to see what issues users were having when dealing with constant subsequences or near constant subsequences.

Yeah... that's a  good idea! I will go and check out the previous issues.

> I think the first part is more important while this second part is "nice to have"

You are right! So, to wrap up, we may do:

* use `T_subseq_isconstant`  in `core._calculate_squared_distance` to make it consistent with `stumpy/stump.py`
* checkout the previous issues regarding identical subseqs
* try setting `config.STUMPY_STDDEV_THRESHOLD` to lower value 

?

Reply:
Yes, that summary sounds good. There is one small note as to why we chose to use the STDDEV to infer a constant region (rather than adding `T_subseq_isconstant`) and that was to save on memory.

Reply:
@seanlaw 
I provided a **recap** in the main page of PR. But, I will leave this conversation open for now just in case you want to mention something.